examples/run_parallel_and_cleanup.py

```diff
@@ -0,0 +1,114 @@
+"""Script to run all example files"""
+import os
+import subprocess
+import shutil
+import sys
+from concurrent.futures import ThreadPoolExecutor, as_completed
+
+#checks if we are in the examples directory
+if not os.path.basename(os.getcwd()) == "examples":
+    print("Run this script from the examples directory.")
+    sys.exit(1)
+# Parse command-line arguments
+if len(sys.argv) > 2:
+    print("Usage: python run_parallel_and_cleanup.py [max_threads]")
+    sys.exit(1)
+
+# Default to 2 threads if no argument is provided
+max_threads = int(sys.argv[1]) if len(sys.argv) == 2 else 4
+if max_threads <= 0:
+    print("Maximum number of threads must be a positive integer.")
+    sys.exit(1)
+
+# Define the log directory
+log_dir = "logs"
+os.makedirs(log_dir, exist_ok=True)
+
+
+def run_script(file_path):
+    """Run a Python script and log its output to a file."""
+    script_name = os.path.basename(file_path)
+    log_file = os.path.join(log_dir, f"{script_name}.log")
+    print(f"Running {file_path}...")
+
+    stdout = ""
+    stderr = ""
+    success = False
+
+    try:
+        # Run the script and capture stdout/stderr
+        result = subprocess.run(["python3", file_path],
+                                capture_output=True,
+                                text=True,
+                                check=True)
+        stdout = result.stdout
+        stderr = result.stderr
+        success = True
+    except subprocess.CalledProcessError as e:
+        stdout = e.stdout
+        stderr = e.stderr
+        success = False
+    finally:
+        # Write logs
+        with open(log_file, "w", encoding="utf-8") as log:
+            log.write(f"--- Output for {script_name} ---\n")
+            log.write("STDOUT:\n")
+            log.write(stdout + "\n")
+            log.write("STDERR:\n")
+            log.write(stderr + "\n")
+
+        status = "\033[92m✔" if success else "\033[91m✘"
+
+        print(f"{status} Logs for {script_name} saved to {log_file}\033[0m")
+
+
+def gather_python_files(path):
+    """Find all Python files in the given directory and its subdirectories.
+    
+    Excludes the current script from the list of files.
+    """
+    current_script = os.path.basename(__file__)
+    python_files = [
+        os.path.join(root, file)
+        for root, _, files in os.walk(path)
+        for file in files
+        if file.endswith(".py") and file != current_script
+    ]
+    return python_files
+
+
+def run_python_files(python_files, max_working_threads):
+    """Run all Python files in parallel with a limit on active threads."""
+    with ThreadPoolExecutor(max_workers=max_working_threads) as executor:
+        futures = [executor.submit(run_script, file) for file in python_files]
```

Comment:
Suggestion: you could instead use ```_= list(executor.map(run_script, python_files))```. The ```list``` ctor processes the results and ensures that any exceptions raised by the threads are raised in the main thread.

Reply:
But what would be the difference?
Is it just readability ?

Reply:
Yes, for readability reasons. And because you are reusing code already provided by the library -- that list comprehension with the ```executor.submit``` method call is used inside the ```executor.map``` method.